specs/distros.rst

```diff
@@ -0,0 +1,580 @@
+API design for the new distros namespace
+========================================
+
+API convention
+--------------
+
+Before dwelving into the details of the proposed API, some conventions
+should be established, so that the API could be pythonic, easy to
+comprehend and extend. We have the following examples of how an object
+should look, depending on its state and behaviour:
+
+ - Use ``.attribute`` if it the attribute is not changeable
+   throughout the life of the object.
+   For instance, the name of a name of a device.
+
+ - Use ``.method()`` for obtaining a variant attribute, which can be
+   different throughout the execution of the object and not modifiable
+   through our object. This is the case for ``device.size()``, we can't
+   set a new size and it can vary throughout the life of the device.
+
+ - For attributes which are modifiable by us and which aren't changing
+   throughout the life of the object, we could use two approaches,
+   a property-based approach or a method based approach.
+
+   The first one is more Pythonic, but can
+   hide from the user of that object the fact that the property does
+   more than it says, the second one is more in-your-face,
+   but feels too much like Java's setters and getters and so on:
+
+       >>> device.mtu
+       1500
+       # actually changing the value, not the cached value.
+       >>> device.mtu = 1400
+       1400
+
+       vs
+
+       >>> device.mtu()
+       1500
+       >>> device.set_mtu(1400)
+       >>>
+
+ - similar to the previous point, we could have ``is_something`` or
+   ``is_something()``. We must choose one of this variants and use it
+   consistently accross the project.
+
+
+Proposed distro hierarchy
+=========================
+
+Both frameworks have a concept of Distro, each different in its way:
+
+    - cloudinit has a ``distros`` location. There is a ``Distro`` base class,
+      with abstract methods implemented by particular distros.
+
+        Problems:
+
+        * not DRY: many implementations have duplicate code with the base class
```

Comment:
This is likely just because we didn't refactor it appropriatel/enough...

I'm not sure this isn't an inherent problem with that way of doing things :wink: